### ISSUE TITLE
Right channel stops playing audio after a few seconds  (samd51)

### ISSUE THREAD

**Issue by jepler** [adafruit/circuitpython#2099]
While trying to reproduce #1992, I ran into this behvaior instead.

 Testing setup: A0, A1 into o'scope.  Put some.wav and code.py on CIRCUITPY.

Initially, A0 and A1 both display a signal, but often A1 stops showing signal before A0 even though both halves of the sample should play for the same duration.  The length of time A1 plays varies.  Oddly, it also varies whether the DAC value is high or low at the end of sample playing, on A0 and A1.

![MSO1104Z_2019-08-31_18 00 40](https://user-images.githubusercontent.com/1517291/64069882-863e7680-cc19-11e9-84c1-4e2977ef62ed.png)
The conjecture is, it's a problem with DMA.

Tested in 4.1.0 and master on a Metro M4 Express.

Code to create wave file: 
```
import array 
from numpy import zeros 
import scipy.io.wavfile 
 
X=-32768+8000 
Y=32767 
 
sz = (512, 2) 
a = zeros(sz, 'h') 
a -= 32768

for i in range(0,sz[0],2):
    a[i,0] = X
    a[i,1] = X
scipy.io.wavfile.write("some.wav", 200000, a)

```

Code to run on circuitpython:
```
import board
import audioio
import array
import time

X = 4000
Y = 65535
data = array.array('H', [0,0,Y,X])
#sample = audioio.RawSample(data, sample_rate=24 * 1000, channel_count=2)
some_wav = open('some.wav', 'rb')
sample = audioio.WaveFile(some_wav)

### A0 will be x, A1 will be y
a = audioio.AudioOut(board.A0, right_channel=board.A1)

while 1:
    a.play(sample, loop=True)
    time.sleep(1)
    a.stop()
    time.sleep(1)

```

**Comment by dhalbert** [comment]
This sounds related to the "refresh" needed on the SAMD51 DAC's. Their output needs to be refreshed explicitly by forcing a reconversion, or else the charge will decay away. If you put various pulldowns on A1 I have a hunch you'll see it decay faster depending on the value of the pulldown.

Turning on refresh is done in `ports/atmel_samd/asf4_conf/hpl_dac_config.h`: `CONF_DAC0_REFRESH` and `CONF_DAC1_REFRESH` are both set to 2. But the symptoms above sound like that value is being changed. Maybe the audio code is somehow ignoring that setting or overriding it?

I originally fixed this issue in https://github.com/adafruit/circuitpython/pull/649. @kevinjwalters mentioned this also in https://github.com/adafruit/circuitpython/issues/1991#issue-466267344.

**Comment by jepler** [comment]
I will check into that, but doesn't this apply only to static signals?  The value is being changed a nominal 100,000 times per second between 0 to 4000 continuously.

**Comment by dhalbert** [comment]
Good point. Is the loss of signal abrupt or does it fade down? Abrupt would sound like a DMA problem (if DMA is being used here).

**Comment by jepler** [comment]
abrupt, like DMA stops

**Comment by kevinjwalters** [comment]
There's also another setting about current (`CONF_DAC0_CCTRL` and `CONF_DAC1_CCTRL`). I've not experimented with this, just came across it in the MicroChip docs. Mentioned on: [Adafruit Forums: What's fastest practical sample_rate for M4 DACs?
](https://forums.adafruit.com/viewtopic.php?f=60&t=153797).

**Comment by jepler** [comment]
I updated the initial comment with new code and an oscope grab.  And just when you got things weird enough, I got  this one too:
![MSO1104Z_2019-08-31_18 10 19](https://user-images.githubusercontent.com/1517291/64069938-cd793700-cc1a-11e9-874b-0bf394b3f829.png)

**Comment by jepler** [comment]
![MSO1104Z_2019-08-31_18 19 51](https://user-images.githubusercontent.com/1517291/64069981-f77f2900-cc1b-11e9-90bf-90a94090778b.png)

**Comment by kevinjwalters** [comment]
I think the SAMD51 has a pseudo-differential mode for the DAC pair? When I saw that in docs I realised bugs around that area could be a cause of _som_e of these coupling-related oddities but there seem to be so many strange quirks that defy an obvious explanation. I did a quick check of the SAMD21 on a CPX a few months back and it's so clean and predictable! It's as if there's a poltergeist in the SAMD51's DACs.

**Comment by kevinjwalters** [comment]
I wonder if we should take [Malvin's advice](https://www.youtube.com/watch?v=GfJJk7i0NTk&t=2m42sl) and start at the beginning? Even at first use the SAMD51 DACs used from CircuitPython behave strangely with #1991 .

**Comment by jepler** [comment]
Based on this looking "right" in an Arduino environment, I am hoping there's just some spot of setup that we've missed.  Unfortunately, I think I've ruled out everything in the DAC's registers.  Moving on to clock setup now, h/t @sommersoft

**Comment by kevinjwalters** [comment]
You are likely to have already observed it or read it, but when I last looked it wasn't clocked fast enough to get the higher rates, 3rd post in https://forums.adafruit.com/viewtopic.php?f=60&t=153797#p759051

**Comment by jepler** [comment]
Since #2552 we use just a single DMA channel to send stereo audio (at least when it's 16 bits).  This should resolve such problems.  If the problem is not resolved with CircuitPython 5.3.0, feel free to reopen this bug.
